File: alhoda/alhoda/report/party_balance_confirmation/party_balance_confirmation.py
# Copyright (c) 2024, Erpgulf and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "label": _("Party Type"),
            "fieldname": "party_type",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Party"),
            "fieldname": "party",
            "fieldtype": "Data",
            "width": 150
        },
        {
            "label": _("Party Name"),
            "fieldname": "party_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("Receivable Account"),
            "fieldname": "party_account",
            "fieldtype": "Link",
            "options": "Account",
            "width": 150
        },
        {
            "label": _("Balance"),
            "fieldname": "balance",
            "fieldtype": "Currency",
            "width": 150
        }
    ]

def get_data(filters):
    conditions = []
    if filters.get("company"):
        conditions.append(f"gl.company = '{filters.get('company')}'")
    if filters.get("party_type"):
        conditions.append(f"gl.party_type = '{filters.get('party_type')}'")
    if filters.get("party"):
        party_list = "', '".join(filters.get("party"))
        conditions.append(f"gl.party IN ('{party_list}')")
    if filters.get("party_account"):
        conditions.append(f"gl.account = '{filters.get('party_account')}'")
    if filters.get("cost_center"):
        conditions.append(f"gl.cost_center = '{filters.get('cost_center')}'")
    if filters.get("report_date"):
        conditions.append(f"gl.posting_date <= '{filters.get('report_date')}'")
    
    where_clause = " AND ".join(conditions)
    
    query = f"""
        SELECT 
            gl.party_type AS party_type,
            gl.party AS party,
            party_name.`{filters.get("party_type")}_name` AS party_name,
            gl.account AS party_account,
            SUM(gl.debit - gl.credit) AS balance
        FROM `tabGL Entry` gl
        LEFT JOIN `tab{filters.get("party_type")}` party_name ON gl.party = party_name.name
        WHERE {where_clause}
        GROUP BY gl.party_type, gl.party, gl.account
        
    """
    data = frappe.db.sql(query, as_dict=True)
    
    # Add formatting for bold if required
    for row in data:
        row["bold"] = row["balance"] < 0  # Example condition to make balance negative bold
    
    return data

File: alhoda/alhoda/report/party_balance_confirmation/party_balance_confirmation.js
// Copyright (c) 2024, Erpgulf and contributors
// For license information, please see license.txt

frappe.query_reports["Party Balance Confirmation"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "report_date",
			label: __("Posting Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
		},

		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
			get_query: () => {
				var company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						company: company,
					},
				};
			},
		},
		{
			fieldname: "party_type",
			label: __("Party Type"),
			fieldtype: "Autocomplete",
			options: Object.keys(frappe.boot.party_account_types),
			on_change: function () {
				frappe.query_report.set_filter_value("party", "");
			},
		},
		{
			fieldname: "party",
			label: __("Party"),
			fieldtype: "MultiSelectList",
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;

				let party_type = frappe.query_report.get_filter_value("party_type");
				if (!party_type) return;

				return frappe.db.get_link_options(party_type, txt);
			},
		},
		{
			fieldname: "party_account",
			label: __("Receivable Account"),
			fieldtype: "Link",
			options: "Account",
			get_query: () => {
				var company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						company: company,
						account_type: "Receivable",
						is_group: 0,
					},
				};
			},
		},
		
		{
			fieldname: "customer_group",
			label: __("Customer Group"),
			fieldtype: "MultiSelectList",
			options: "Customer Group",
			get_data: function (txt) {
				return frappe.db.get_link_options("Customer Group", txt);
			},
		},

		{
			fieldname: "email_res1",
			label: __("Email REsponse 1"),
			fieldtype: "Data",
			default: "Syed Imran (Syedimran@kpmg.com)",


		},
		{
			fieldname: "email_res2",
			label: __("Email REsponse 2"),
			fieldtype: "Data",
			default: "Usama Abdulsamad (Uasamad@kpmg.com)",


		}, 
		{
			fieldname: "email_auditor",
			label: __("Email Auditor"),
			fieldtype: "Data",
			default: "audit_confirmation_2022@stayromax.com",


		},


	
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (data && data.bold) {
			value = value.bold();
		}
		return value;
	},

	
};


function get_party_type_options() {
	let options = [];
	frappe.db
		.get_list("Party Type", { filters: { account_type: "Receivable" }, fields: ["name"] })
		.then((res) => {
			res.forEach((party_type) => {
				options.push(party_type.name);
			});
		});
	return options;
}